src/libraries/System.Text.Json/gen/Reflection/ReflectionExtensions.cs

```diff
@@ -21,12 +22,7 @@ public static TValue GetConstructorArgument<TValue>(this CustomAttributeData cus
         public static bool IsInitOnly(this MethodInfo method)
         {
             MethodInfoWrapper? methodInfoWrapper = method as MethodInfoWrapper;
-
-            if (methodInfoWrapper == null)
-            {
-                throw new ArgumentException("Expected a MethodInfoWrapper instance.", nameof(method));
-            }
-
+            Debug.Assert(methodInfoWrapper != null);
```

Comment:
Seems like a straight cast would be better here.

```C#
MethodInfoWrapper methodInfoWrapper = (MethodInfoWrapper)method;
```

Then when it fails, you get an InvalidCastException, and it is easier to debug than either an assertion failing, or even worse in Release getting a NRE.